### PR TITLE
Add `jsonc` parser

### LINKED ISSUE #15553: Prettier should format JSONC with trailing commas
- Right now, Prettier's default value of "trailingCommas" is "all", which is fantastic. Presumably, most people who use Prettier would agree with this, as the reasons for wanting trailing commas are compelling: better Git diffs, easier refactoring, and so on.
- For all the same reasons that we would want trailing commas in a JavaScript/TypeScript file, we also want to have trailing commas in JSONC files.
- However, Prettier does not insert trailing commas into JSONC files by default.

You could frame this problem as either:
1. a bug in Prettier (since the default value of "trailingCommas" is supposed to be "all"), or
2. a failure in Prettier parsing JSONC properly

Either way, this should be changed.

<br>

## Steps to Reproduce

```ts
mkdir test
cd test
npm init --yes
npm install prettier --save-dev
echo '{ "foo": "bar" }' > test.jsonc
npx prettier test.jsonc --write
cat test.jsonc
```

Observe that the formatted JSONC file does not have trailing commas.

<br>

## Current Hacky Workaround

Use the following custom Prettier config:

```
  overrides: [
    {
      files: ["**/*.jsonc"],
      options: {
        parser: "json5",
        quoteProps: "preserve",
      },
    },
```

Since JSONC is a subset of JSON5, this works properly. However, it is important that we also specify `quoteProps: "preserve"`, as the default value for `quoteProps` for JSON5 is "as-needed", which would unquote JSON properties, making it non-valid JSONC.

Obviously, having to use a custom Prettier config is really unfortunate, as one of the main ideas of using Prettier is that you shouldn't use your own custom config, you should just use all the default values. But this bug / bad default behavior prevents that.

Also, as a semi-related note, it was proposed that the default value for `quoteProps` in JSON5 would be changed in #4639. If a PR for that was accepted, then we could simplify the workaround slightly, although the ideal solution is to not to have to do a workaround to begin with.

<br>

## Summarizing Prior Discussion

This issue continues #5708.

The core issue is that Prettier does not have a separate parser for JSONC files, it just uses the JSON parser. Since the JSON parser works with comments, this "just works" some of the time, but crashes and burns when trailing commas are involved.

ikatyang says [here](https://github.com/prettier/prettier/issues/5708#issuecomment-452523700) that they considered making a JSONC parser, but this was confusing, as it was really just a trailing commas parser and not a comment parser. Which is definitely a valid concern!

However, five years have passed since that comment, and JSONC files have only become more prevalent in the ecosystem. For example, popular tools like [CSpell](https://github.com/streetsidesoftware/cspell) (code spell checker), [Knip](https://github.com/webpro/knip) (dead code finder), and [TypeDoc](https://github.com/TypeStrong/typedoc) (documentation generator) all support JSONC files by default ("cspell.jsonc", "knip.jsonc", and "typedoc.jsonc" respectively).

Thus, there is a more compelling reason for Prettier to support the JSONC format more robustly. To @ikatyang, and the rest of the Prettier team, can we revisit this decision? A dedicated parser for JSONC means that we can finally format JSONC files properly out of the box without having to use the dirty JSON5 hack specified above.

### PR DESCRIPTION
**Pull request by fisker**
## Description

Fixes #15553

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: docs/options.md [hunk 1]
```diff
@@ -115,6 +115,7 @@ Note that Prettier never unquotes numeric property names in Angular expressions,
 [quote-props-flow]: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBjOA7AzgFzAA8wBeMAb1TDAAYAuMARlQF8g
 [quote-props-vue]: https://github.com/prettier/prettier/issues/10127
 
+<!-- TODO[@fisker]: Check this -->
 If this option is set to `preserve`, `singleQuote` to `false` (default value), and `parser` to `json5`, double quotes are always used for strings. This effectively allows using the `json5` parser for “JSON with comments and trailing commas”.
```
Comment (fisker) [inline docs/options.md:119]: @thorn0 Do you think I can revert #10323?

File: src/language-json/languages.evaluate.js [hunk 2]
```diff
@@ -15,10 +15,9 @@ const languages = [
     vscodeLanguageIds: ["json"],
     extensions: data.extensions.filter((extension) => extension !== ".jsonl"),
   })),
-  createLanguage(linguistLanguages["JSON with Comments"], (data) => ({
-    parsers: ["json"],
+  createLanguage(linguistLanguages["JSON with Comments"], () => ({
+    parsers: ["jsonc"],
     vscodeLanguageIds: ["jsonc"],
-    filenames: [...data.filenames, ".eslintrc"],
```
Comment (fisker) [inline src/language-json/languages.evaluate.js:21]: `.eslintrc` is not listed in https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file-formats, and `.eslintrc.json` already in `linguistLanguages`.

File: docs/options.md (hunk 1 above)
Comment (thorn0) [reply docs/options.md]: It'd be a breaking change with little gain. Maybe better to keep the logic and just remove this part from the docs.

Comment (fisker) [reply docs/options.md]: https://github.com/prettier/prettier/pull/15831/commits/ed020b28cacfd48982b313e4a36e5caa042cd664

**Review by sosukesuzuki — APPROVED** [review]

**Pull request merged**